### PR TITLE
[MXNET-314] Support Integer Type parsing in CSVIter

### PR DESCRIPTION
**Pull request by haojin2**
## Description ##
Support parsing of int32 type in CSVIter

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-314]
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Support integer data type in CSVIter

## Comments ##
In the past all parsers default to parse the files in real_t type, but when the files contain large integers we lose precision. e.g. 20000001 actually will have the same float32 representation as 20000000. To provide a way to accurately read the data for users who have such needs, we need to let the iterators be able to support parsing the input texts as integers.
This PR depends on dmlc/dmlc-core#388 to build work properly.
Comes together with a fix for flaky test_correlation

### CODE REVIEW

File: src/io/iter_csv.cc [hunk 1]
```diff
@@ -88,6 +154,34 @@ class CSVIter: public IIterator<DataInst> {
...
+  std::unique_ptr<dmlc::Parser<uint32_t, int> > label_parser_;
+  std::unique_ptr<dmlc::Parser<uint32_t, int> > data_parser_;
+};
+
+class CSVIterFloat: public CSVIterBase {
+ public:
+  virtual ~CSVIterFloat() {}
+  virtual void Init(const std::vector<std::pair<std::string, std::string> >& kwargs) {
+    param_.InitAllowUnknown(kwargs);
+    data_parser_.reset(dmlc::Parser<uint32_t>::Create(param_.data_csv.c_str(), 0, 1, "csv"));
```
[the remark this answers is not in the document]
Comment (haojin2) [reply src/io/iter_csv.cc]: Done

**Comment by haojin2** [comment]
@rahul003 @anirudh2290 @reminisce @eric-haibin-lin Please review if you have any time, as this has been pending for a long time. Thanks!

File: tests/python/unittest/test_io.py [hunk 2]
```diff
@@ -294,7 +294,7 @@ def test_DataBatch():
 
 @unittest.skip("test fails intermittently. temporarily disabled till it gets fixed. tracked at https://github.com/apache/incubator-mxnet/issues/7826")
```
Comment (anirudh2290) [inline tests/python/unittest/test_io.py]: seems like this test is being skipped . is the issue fixed now ?

File: tests/python/unittest/test_io.py [hunk 3]
```diff
@@ -306,12 +306,13 @@ def check_CSVIter_synthetic():
                 fout.write('0\n')
 
         data_train = mx.io.CSVIter(data_csv=data_path, data_shape=(8,8),
-                                   label_csv=label_path, batch_size=100)
-        expected = mx.nd.ones((100, 8, 8))
+                                   label_csv=label_path, batch_size=100, dtype=dtype)
+        expected = mx.nd.ones((100, 8, 8), dtype=dtype)
         for batch in iter(data_train):
             assert_almost_equal(data_train.getdata().asnumpy(), expected.asnumpy())
```
Comment (anirudh2290) [inline tests/python/unittest/test_io.py]: we may need to set some atol and rtol

File: src/io/iter_csv.cc [hunk 4]
```diff
@@ -54,27 +57,90 @@ struct CSVIterParam : public dmlc::Parameter<CSVIterParam> {
...
+  unsigned inst_counter_{0};
+  // at end
+  bool end_{false};
 
+  // label parser
+  size_t label_ptr_{0}, label_size_{0};
+  size_t data_ptr_{0}, data_size_{0};
+};
+
+class CSVIterInt: public CSVIterBase {
```
Comment (anirudh2290) [inline src/io/iter_csv.cc]: Seem like, the only difference between CSVIterInt and CSVIterFloat is the data  type for the Parser. Can this code duplication be avoided ?

File: tests/python/unittest/test_io.py (hunk 3 above)
Comment (haojin2) [reply tests/python/unittest/test_io.py]: We need the exact result from parsing, the fundamental cause for this PR was that using float32 for parsing large integers is inaccurate.

File: tests/python/unittest/test_io.py (hunk 2 above)
Comment (haojin2) [reply tests/python/unittest/test_io.py]: I looked at the issue mentioned in the skip message, it was some issue on Windows. I've manually run this test on my Ubuntu machine and this test passes for both dtypes. On the other hand this PR has already been tested on real workloads and proven to work.

File: src/io/image_iter_common.h [hunk 5]
```diff
@@ -337,21 +337,11 @@ struct ImageDetNormalizeParam :  public dmlc::Parameter<ImageDetNormalizeParam>
...
   /*! \brief number of prefetched batches */
   size_t prefetch_buffer;
-  /*! \brief data type */
-  dmlc::optional<int> dtype;
 
   // declare parameters
   DMLC_DECLARE_PARAMETER(PrefetcherParam) {
     DMLC_DECLARE_FIELD(prefetch_buffer).set_default(4)
         .describe("Maximum number of batches to prefetch.");
-    DMLC_DECLARE_FIELD(dtype)
```
Comment (eric-haibin-lin) [inline src/io/image_iter_common.h]: dtype is problematic for PrefetcherIter but I'm afraid this change will break API. If someone use an data iter with dtype = float32, it will throw an exception.

File: src/io/iter_csv.cc (hunk 4 above)
Comment (eric-haibin-lin) [reply src/io/iter_csv.cc]: Can we just make CSVIter templated?

Comment (haojin2) [reply src/io/iter_csv.cc]: This may not be possible as this is a user-facing class.

File: tests/python/unittest/test_io.py (hunk 3 above)
Comment (eric-haibin-lin) [reply tests/python/unittest/test_io.py]: Does this test actually check the dtype? assert_almost_equal only checks the value but not the dtype
Suggest test:
1. assert(data_train.getdata() == dtype)
2. When dtype is int32, generate large int like 50000001 and test if the data loaded is exactly 50000001.

File: src/io/iter_csv.cc [hunk 6]
```diff
@@ -116,33 +147,59 @@ class CSVIter: public IIterator<DataInst> {
...
+
+  // intialize iterator loads data in
+  virtual void Init(const std::vector<std::pair<std::string, std::string> >& kwargs) {
+    param_.InitAllowUnknown(kwargs);
+    bool dtype_has_value = false;
+    int target_dtype = -1;
+    for (const auto& arg : kwargs) {
+      if (arg.first == "dtype") {
+        dtype_has_value = true;
+        target_dtype = (arg.second == "int32") ? mshadow::kInt32 : mshadow::kFloat32;
```
Comment (eric-haibin-lin) [inline src/io/iter_csv.cc]: The user won't be aware that only int32 and float32 are supported. A warning should be printed (using `LogOnce`) if any unsupported dtype is used.

File: src/io/iter_csv.cc [hunk 7]
```diff
@@ -116,33 +147,59 @@ class CSVIter: public IIterator<DataInst> {
...
-  bool end_{false};
-  // dummy label
-  mshadow::TensorContainer<cpu, 1, real_t> dummy_label;
-  // label parser
-  size_t label_ptr_{0}, label_size_{0};
-  size_t data_ptr_{0}, data_size_{0};
-  std::unique_ptr<dmlc::Parser<uint32_t> > label_parser_;
-  std::unique_ptr<dmlc::Parser<uint32_t> > data_parser_;
+  std::unique_ptr<CSVIterBase> iterator_;
 };
```
Comment (eric-haibin-lin) [inline src/io/iter_csv.cc]: Also need to update the documentation that dtype only supports int32 and fp32.

File: tests/python/unittest/test_io.py (hunk 3 above)
Comment (eric-haibin-lin) [reply tests/python/unittest/test_io.py]: Even if you don't implement int32 parser this test will pass:
```
>>> a = mx.nd.ones((2,2), dtype='int32')
>>> b = mx.nd.ones((2,2), dtype='float32')
>>> mx.test_utils.assert_almost_equal(a.asnumpy(), b.asnumpy())
>>> # no exception
```

File: src/io/iter_csv.cc (hunk 7 above)
Comment (haojin2) [reply src/io/iter_csv.cc]: Done

File: src/io/iter_csv.cc (hunk 6 above)
Comment (haojin2) [reply src/io/iter_csv.cc]: Maybe we should use CHECK here? cause if the dtype is unsupported then the consequent training steps will also fail.

File: tests/python/unittest/test_operator.py [hunk 8]
```diff
@@ -2491,7 +2491,7 @@ def test_infer_type(dtype):
                                                 names=['a', 'b'])
             raise AssertionError(msg)
 
-    for dtype in ['float16', 'float32', 'float64']:
+    for dtype in ['float16', 'float32']:
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_operator.py]: Please track https://github.com/apache/incubator-mxnet/issues/10280 ...
